### PR TITLE
add contract tasks to hardhat

### PR DESCRIPTION
**Pull request by jaycenhorton**
### Changes
This adds the ability to call contract functions from hardhat (via the cli as well as via in-code via the hardhat runtime)

#### Examples

Note that the below are just two examples, but these tasks are generalized and should be able to call any contract tu

##### mint Nori_V0

```bash
hardhat NORI --func mint  "0x321af43416f670ce8b4ba214dfb87c4199e2a77f" 1000000000000000000 --network mumbai
```

##### mint NORI

```bash
hardhat Nori_V0 --func mint  "0x321af43416f670ce8b4ba214dfb87c4199e2a77f" 1000000000000000000 "0x" "0x"  --network mumbai
```

---
Note that these tasks can also be called from within code which might be useful for our cypress/integration test

```ts
await hre.run('NORI', {
  func: 'mint',
  args: [
    '0x321af43416f670ce8b4ba214dfb87c4199e2a77f',
    1000000000000000000,
    '0x0',
    '0x0',
  ],
});
```

### CODE REVIEW

**Review by scottynomad — APPROVED** [review]

File: tasks/nori-v0.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as noriV0Abi } from '@/artifacts/Nori_V0.sol/Nori_V0.json';
+
+export const TASK = {
+  name: 'Nori_V0',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/nori-v0.ts:13]:
```suggestion
  description: 'Interact with the Nori_V0 contract',
```

File: tasks/nccr-v0.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as nccrAbi } from '@/artifacts/NCCR_V0.sol/NCCR_V0.json';
+
+export const TASK = {
+  name: 'NCCR_V0',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/nccr-v0.ts:13]:
```suggestion
  description: 'Interact with the NCCR_V0 contract',
```

File: tasks/fifo-market.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as fifoMarketAbi } from '@/artifacts/FIFOMarket.sol/FIFOMarket.json';
+
+export const TASK = {
+  name: 'FIFOMarket',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/fifo-market.ts:13]:
```suggestion
  description: 'Interact with the FIFOMarket contract',
```

File: tasks/certificate.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as certificateAbi } from '@/artifacts/Certificate.sol/Certificate.json';
+
+export const TASK = {
+  name: 'Certificate',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/certificate.ts:13]:
```suggestion
  description: 'Interact with the Certificate contract',
```

File: tasks/nori.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as noriAbi } from '@/artifacts/NORI.sol/NORI.json';
+
+export const TASK = {
+  name: 'NORI',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/nori.ts:13]:
```suggestion
  description: 'Interact with the NORI contract',
```

File: tasks/removal.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  CONTRACT_FUNCTION_TASK_PARAMETERS,
+  CONTRACT_FUNCTION_TASK_RUN,
+} from './utils/contract-functions';
+
+import * as contractsConfig from '@/contracts.json';
+import { abi as removalAbi } from '@/artifacts/Removal.sol/Removal.json';
+
+export const TASK = {
+  name: 'Removal',
+  description: 'Interact with the nori contract',
```
Comment (amiecorso) [inline tasks/removal.ts:13]:
```suggestion
  description: 'Interact with the Removal contract',
```

File: types/global.d.ts [hunk 7]
```diff
@@ -111,4 +99,15 @@ declare global {
...
     (hre: CustomHardHatRuntimeEnvironment): Promise<void | boolean>;
   }
+
+  namespace NodeJS {
+    interface ProcessEnv {
+      MNEMONIC?: string;
+      STAGING_MNEMONIC?: string;
+      INFURA_STAGING_KEY?: string;
+      TENDERLY_USERNAME?: string;
+      TENDERLY_PROJECT?: string;
```
Comment (amiecorso) [inline types/global.d.ts:109]:
```suggestion
      TENDERLY_PROJECT?: string;
      ETHERNAL_EMAIL?: string;
      ETHERNAL_PASSWORD?: string;
```

File: tasks/utils/contract-functions.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  contractAbi: ConstructorParameters<typeof Contract>[1];
+  from: number;
+  func: string;
+  args: unknown[];
+  hre: CustomHardHatRuntimeEnvironment;
+}): Promise<void> => {
+  if (hre.network.provider) {
+    const signers = await hre.ethers.getSigners();
+    const signer = signers[from];
+    const noriV0TokenContract = new Contract(
```
Comment (amiecorso) [inline tasks/utils/contract-functions.ts:43]:
```suggestion
    const contract = new Contract(
```

File: tasks/utils/contract-functions.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      contractAddress,
+      contractAbi, // todo store artifacts in contracts.json
+      signer
+    );
+    const transaction = await noriV0TokenContract[func](...args);
```
Comment (amiecorso) [inline tasks/utils/contract-functions.ts:48]:
```suggestion
    const transaction = await contract[func](...args);
```

**Review by amiecorso — APPROVED** [review]
awesome!
approving but notice the suggestions

**Pull request merged**